### ISSUE TITLE
Resolve performance issues

### ISSUE THREAD

**Issue by chipsenkbeil** [chipsenkbeil/vimwiki-rs#23]
Having written `SpanFactory` and leveraging other hacks on top of `nom_locate`, the performance of reading a file has massively dropped. For the pandoc example, the performance prior to incorporating `SpanFactory` was 1.5s (already not good) and post incorporation is 30~35s!

My current thoughts are:
1. Parser combinators do not require the input to be copy. Rather, they take in the input and provide it back, which I use. Try to avoid a copy with our hack struct as my guess is copying around the struct is causing massive slowdowns
    a. We need to update `scan(...)` to not consume the input
2. Remove `nom_locate` in favor of writing my own span struct. To make things easier (and more efficient), leverage the [`Bytes`](https://crates.io/crates/bytes) crate. This crate maintains multiple reference counts and allows overlapping slices, etc. Main draw for me is the ability to chain non-consecutive slices. Have the wrapping struct containing an instance of `Bytes` implement all traits needed by `nom` input and NOT be a copy to see how well it performs.
    a. With the struct, we would maintain a master `Bytes` reference and the shortened `Bytes` chain based on removed slices, just like we are now with the hack approach.
    b. We would re-implement the `get_utf8_column` without the need for unsafe as we can leverage the master `Bytes` instance to get the proper start (not needing a consecutive slice). `nom_locate` uses [`bytecount`](https://crates.io/crates/bytecount), which counts over a byte slice. If needed, I'm assuming I could count of multiple byte slices and then sum up in the same way.

**Comment by chipsenkbeil** [comment]
## Prior to supporting segmented spans using copy approach

![image](https://user-images.githubusercontent.com/2481802/93687963-de942080-fa87-11ea-89d4-57fd1b5a2b8e.png)

## Supporting segmented spans using SpanFactory & copy approach

![image](https://user-images.githubusercontent.com/2481802/93687968-e653c500-fa87-11ea-8f73-3fa035fd7cb2.png)

## Supporting segmented spans using Bytes & clone approach

![image](https://user-images.githubusercontent.com/2481802/93687975-ee136980-fa87-11ea-9d3a-d49005d6eded.png)

**Comment by chipsenkbeil** [comment]
I'm sure there are further optimizations that can be made such as fiddling with the caching attempt that was made to see if it really helps or if the cost of the clones slows things down.

### Span Optimizations

1. Could calculate column changes on slice using `enumerate` on the line iter, taking the last result (with enumerate telling us lines - 1) and using the position as starting point for `bytescount::num_chr`
2. Could remove the `Arc<Mutex<Option<usize>>` for various fields to see if the cost of cloning is too great
3. Could continue to experiment with slice segment, seeing if a chain of byte slices that has NOT been converted to a single `Bytes` instance would help (cost of mem copy... would be really hard and I don't think this is a problem)
4. Precalculate the line offset positions and cache them for the entire span using memchr; would reduce needing to search for lines on slice to determine consumed lines, current line, etc. (**this might be a good one**)

### Parser Optimizations

Need to evaluate the parsers a bit more to see what we can improve. `inline_component` and `inline_component_container` seem like the possible good choices if there's anything to do there.

**Comment by chipsenkbeil** [comment]
Another minor update is that the [`specification.wiki`](https://github.com/vimwiki/vimwiki/pull/1025) file I wrote for the vimwiki project takes ~35 minutes to parse! Wowzer! This will be a good marker for parsing improvement. Shooting for an order of magnitude improvement should be the first goal.

**Comment by chipsenkbeil** [comment]
Pandoc is able to parse `specification.wiki` and produce `specification.html` via `pandoc -f vimwiki -t html specification.wiki` in 1.381s according to `time pandoc -f vimwiki -t html specification.wiki`.

Need to review how the pandoc parser is written to see if there are poor choices I've made that can be corrected to speed things up.

**Comment by chipsenkbeil** [comment]
Another update is that when building the release version of our server, the parsing time drops by an order of magnitude to somewhere in the ~3-4 minute range. This is obviously a bit of a relief, but we still need to achieve at least another order of magnitude to be in the realm of pandoc and, honestly, I would expect even better out of Rust.

A couple of thoughts:
1. Be smarter for block-level parsing by grouping elements with similar starting characters. For instance, blank lines, lists, tables, headers, and paragraphs (if we support 1-3 spaces) all can start with some amount of spaces. Rather than checking the same spaces repeatedly, look for `space0` and then try each in turn.
2. Pandoc has some short-circuiting for paragraphs and a couple of other elements, that may be something we can copy for here.
3. Per earlier comments, try leveraging `memchr` to find surround characters for inline processing such as `*` and `_`. Use `memchr2` with the character and a newline so we can quickly determine if the character occurs before the next line, unless we want to pre-calculate all newline positions in our buffer beforehand.
4. Also mentioned above, extract inline elements like code to not parse any other inline elements inside.
5. Looking closer at inline elements, should refactor decorated text to just be an enum of the types rather than its convoluted setup. Assume that we won't support nested types like bold strikeout. This means that we won't need to parse decorated text within decorated text.

**Comment by chipsenkbeil** [comment]
`memchr` does indeed help. Other factors that have improved performance noticeably:
1. Not calculating line/column - this is the biggest drain on performance and, having realized vim can report your cursor as a byte offset, it is no longer needed to be calculated
2. Switching away from custom spans that calculated above line/column on slice
3. Skipping inline parsers within text parser until we see one of several special starting characters

**Comment by chipsenkbeil** [comment]
Turns out, the error reporting was also immediately calling line() and column() on every error, which is quite often given we try parsers quite a lot. When switching back to my old branch with the lazy region and commenting out the error line/column calcuations, the speed was on par with my other attempt with offset and len.

I like parts of each and vim would support each. I think line/column gives more detail for humans. Offset would be more useful for machines (moving some byte span, etc).

Other aspect is if calculating located elements is too expensive, or if it was actually the error type that was the biggest problem.

**Comment by chipsenkbeil** [comment]
Final notes. I compared the two branches I had after adjusting errors to NOT calculate line/column eagerly.
1. Region is just offset + len: ~2.66s for `specification.wiki` with debug build
2. Region eagerly calculates line & column when parsing elements (but not errors): ~2.74s for `specification.wiki` with debug build

I decided to pursue branch 1 as I can still calculate line & column position on demand as long as the file itself is still accessible. All other work I would do such as moving elements around can be done by offset + len and calculating line & column when required. This means a higher cost per operation, but a lower cost for parsing, which I think is a fine tradeoff.

One thought for the future would be to provide an `Option<LineColumn>` within `Region` that can be used to store line & column information (start and end) if we eagerly want it. This would enable us to unlock the advantage of branch 2, make it configurable, and not deal with lifetimes for `Located<...>` and `Region`.

Provide some method like `fn with_line_column(self, input: Span) -> Self` for region that would compute the line and column information and produce a new region with that information encoded.
